### PR TITLE
RDS & Redshfit support for AWS China regions (part 1?)

### LINKED ISSUE #10540: Database Access for China RDS improvements
### What

Two improvements should be made to Database Access to better support RDS databases in China regions (e.g. `cn-north-1`):

- [ ] Automatic CA downloader should be updated to fetch CA bundles specific to Chinese regions: https://docs.amazonaws.cn/en_us/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html.
- [ ] Code that tries to parse database URI to determine whether it's an RDS instance should recognize `.amazonaws.com.cn` suffix: https://github.com/gravitational/teleport/blob/v8.3.1/api/types/database.go#L359-L369.

### PR DESCRIPTION
**Pull request by greedy52**
Implements #10540 
- #10540

## changes

- adds new package `api/utils/aws` and moves some existing functions there
- adds support for RDS/Redshift url parsing for AWS China regions
- adds support for RDS/Redshift CA urls for AWS China regions
- UTs

## Testing
- [x] existing UT passes (no existing logic breaks)
- [x] test RDS/Redshift in ca-central-1
- [ ] test RDS/Redshift in cn-north-1 ?

(note, I marked title as part 1, since not sure what else needs to change as we don't have access to real China reginos for testing)

### CODE REVIEW

File: api/utils/aws/endpoint.go
Comment (jakule) [inline api/utils/aws/endpoint.go:109]: I'm not a huge fan of regex, but in this case I think that something like `.*\.([\w\d-]+.rds|rds.[\w\d-]+).amazonaws.com.*` looks cleaner.

File: api/utils/aws/endpoint.go
Comment (smallinsky) [inline api/utils/aws/endpoint.go:52]:
```suggestion
		return strings.TrimSuffix(endpoint, AWSEndpointSuffix), nil
```

File: api/utils/aws/endpoint.go
Comment (smallinsky) [inline api/utils/aws/endpoint.go:56]:
```suggestion
		return strings.TrimSuffix(endpoint, AWSCNEndpointSuffix), nil
```

File: api/utils/aws/endpoint.go
Comment (smallinsky) [inline api/utils/aws/endpoint.go:59]:
```suggestion
	return "", trace.BadParameter("endpoint %v is not an AWS endpoint", endpoint)
```

**Review by smallinsky — COMMENTED** [review]
Left some minor comments. The changes LGTM but will hold with approval till real ` cn-north-1`rds instance test.

File: api/utils/aws/endpoint.go
Comment (greedy52) [reply api/utils/aws/endpoint.go:109]: i don't like regex either in general either but it may not be a bad idea in this case. let me try it out. there are other partitions we don't support yet that actually has suffix like `.sc2s.sgov.gov` but those can be added to regex too.

Comment (greedy52) [reply api/utils/aws/endpoint.go:109]:
```
var redshiftRegex = regexp.MustCompile(`^(?P<id>[\w\d-]+)\.(?:[\w\d-]+)\.(?:redshift\.(?P<region>[\w\d-]+)|(?P<region>[\w\d-]+)\.redshift)\.amazonaws.com(?:\.cn)?$`)
                                                                                                                                  
func ParseRedshiftURI(uri string) (clusterID, region string, err error) {                                                        
    endpoint, _, err := net.SplitHostPort(uri)                                                                                    
    if err != nil {                                                                                                               
        return "", "", trace.Wrap(err)                                                                                            
    }                                                                                                                             
                                                                                                                                  
    values := redshiftRegex.FindStringSubmatch(endpoint)                                                                          
    if len(values) != 4 {                                                                                                         
        return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", uri)                                         
    }                                                                                                                             
                                                                                                                                  
    // values[0] is the capture of the entire string.
    // values[1] is <id>.
    // values[2] and values[3] is <region> depends on the position.
    return values[1], (values[2] + values[3]), nil

}
```

This is the best i can do with my limited regex knowledge. The regex gets complicated with capture groups. not sure if this reads better than current version.

WDYT?

Comment (jakule) [reply api/utils/aws/endpoint.go:109]: In my opinion this is cleaner than the current implementation with many `split` and boundy checks. I'd just add a comment with all strings that should match that regex or even a UT to make it clear what we want to match.

@r0mant @smallinsky Any opinions?

Comment (smallinsky) [reply api/utils/aws/endpoint.go:109]:
```go
       parts := strings.Split(trimmedEndpoint, ".")
	if len(parts) != 4 {
		return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", endpoint)
	}

	// Service name/region can be at either position 2 or 3.
	switch {
	case parts[3] == RedshiftServiceName:
		region = parts[2]

	case parts[2] == RedshiftServiceName:
		region = parts[3]

	default:
		return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", endpoint)
	}

	return parts[0], region, nil
```
vs 
```go
    var redshiftRegex = regexp.MustCompile(`^(?P<id>[\w\d-]+)\.(?:[\w\d-]+)\.(?:redshift\.(?P<region>[\w\d-]+)|(?P<region>[\w\d-]+)\.redshift)\.amazonaws.com(?:\.cn)?$`)
 ...
    values := redshiftRegex.FindStringSubmatch(endpoint)                                                                          
    if len(values) != 4 {                                                                                                         
        return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", uri)                                         
    }                                                                                                                             
                                                                                                                                  
    // values[0] is the capture of the entire string.
    // values[1] is <id>.
    // values[2] and values[3] is <region> depends on the position.
    return values[1], (values[2] + values[3]), nil
```

Personally I would incline to the manual parsing version. Regex for me is like blackbox and really hard to understand only by reading but this is only my preference.

File: api/utils/aws/endpoint.go
Comment (r0mant) [inline api/utils/aws/endpoint.go:94]: I wonder if the function's logic was cleaner if you split it in specific methods for parsing regular and cn endpoints e.g.

```go
func ParseRDSEndpoint(endpoint string) (id, region string, err error) {
  if strings.HasSuffix(endpoint, AWSCNEndpointSuffix) {
    return parseRDSCNEndpoint(endpoint)
  }
  return parseRDSEndpoint(endpoint)
}

func parseRDSEndpoint(endpoint string) (id, region string, err error) {
  //
}

func parseRDSCNEndpoint(endpoint string) (id, region string, err error) {
  //
}
```

Then this logic can be extended quite easily and cleanly to other types of endpoints (if needed).

File: api/utils/aws/endpoint.go
Comment (r0mant) [reply api/utils/aws/endpoint.go:109]: I think that manual parsing is easier to understand also, but I would consider splitting this into `parseRedshiftURI` and `parseRedshiftCNURI` similar to my RDS suggestion above which would make it a little more cleaner (I think).

File: api/utils/aws/endpoint.go
Comment (r0mant) [inline api/utils/aws/endpoint.go:77]: This function is called ParseRDS**Endpoint** but the Redshift one is called ParseRedshift**URI**. I would stick to consistent terminology.

File: api/utils/aws/endpoint.go
Comment (r0mant) [inline api/utils/aws/endpoint.go:63]: Let's choose "Endpoint" vs "URI" and use is consistently for these methods.

File: api/utils/aws/endpoint.go
Comment (greedy52) [reply api/utils/aws/endpoint.go:77]: For RDS there are two functions `ParseRDSURI` vs `ParseRDSEndpoint`.

URI is passed from api/types with the port. I feel the port is mandatory in that case. 
But there is another place we need RDS endpoint parsing but only have host name without the port. 

Maybe I can do the same for redshift to have two functions?  or maybe I can keep the `ParseRDSEndpoint` version, and do the `SplitHostPort` outside?

What do you recommend?

File: api/utils/aws/endpoint.go
Comment (greedy52) [reply api/utils/aws/endpoint.go:109]: I have an upcoming change for RDS proxy which will make this parsing even messier. And another potential change is the redshift serverless

How about something like this
```
var rdsEndpointFormats = []string{                                               
    "<id>.<hash>.<region>.rds.amazonaws.com",                                       
    "<id>.<hash>.rds.<region>.amazonaws.com.cn",                                    
}                                
                                                                                                       
func ParseRDSEndpoint(endpoint string) (id, region string, err error) {                                
    for _, format := range rdsEndpointFormats {                                
        if values, ok := parseEndpointWithFormat(endpoint, format); ok {                       
            return values["id"], values["region"], nil                                
        }                                             
    }                                                 
    return "", "", trace.BadParameter("failed to parse %v as RDS endpoint", endpoint)                                
}  
```
where format is a custom format that has keys surrounded by `<>`

I have implemented `parseEndpointWithFormat` locally, which a small function (about 30lines) that translate this format to a regex for matching and capturing the values.

Obviously performance won't be as good as manual parsing, but it's easier to add more formats later.

What do you guys think? @r0mant @smallinsky @jakule

File: api/utils/aws/endpoint.go
Comment (greedy52) [reply api/utils/aws/endpoint.go]: Decided to have a single function `ParseRDSEndpoint` and only do the `SplitHostPort` when `:` is found.

File: api/utils/aws/endpoint.go
Comment (r0mant) [inline api/utils/aws/endpoint.go:128]:
```suggestion
		return "", "", trace.BadParameter("failed to parse %v as Redshift CN endpoint", endpoint)
```

File: api/utils/aws/endpoint.go
Comment (r0mant) [inline api/utils/aws/endpoint.go:86]:
```suggestion
		return "", "", trace.BadParameter("failed to parse %v as RDS CN endpoint", endpoint)
```

**Review by r0mant — APPROVED** [review]
@greedy52 Please backport to v9 and v8 also.

**Review by smallinsky — APPROVED** [review]

**Review by jakule — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Comment by r0mant** [comment]
@smallinsky @greedy52 Are we blocked on getting access to AWS China account or are we reasonably sure the changes will work and can merge this?

**Comment by greedy52** [comment]
@r0mant I am confident on what's changed in this PR (CA, and URL parsing). Just not sure if there will be other issues that breaks RDS/Redshift in China. I'd love to merge this to master as it's blocking my other changes. I can test in China region when it's availble.

**Comment by r0mant** [comment]
@greedy52 Sounds good, let's merge this and remember to test once we get AWS China access (we're working on it).

**Pull request merged**
